storm-core/src/clj/backtype/storm/daemon/supervisor.clj

```diff
@@ -372,66 +389,50 @@
-            stormroot (supervisor-stormdist-root conf storm-id)]
-        (if-not (or (contains? downloaded-storm-ids storm-id) (.exists (File. stormroot)) (nil? master-code-dir))
-          (download-storm-code conf storm-id master-code-dir supervisor download-lock))
         ))
-
-    (wait-for-workers-launch
-     conf
-     (dofor [[port assignment] reassign-executors]
-            (let [id (new-worker-ids port)
-                  storm-id (:storm-id assignment)
-                  ^WorkerResources resources (:resources assignment)
-                  mem-onheap (.get_mem_on_heap resources)]
-              (try
-                (log-message "Launching worker with assignment "
-                             (pr-str assignment)
-                             " for this supervisor "
-                             (:supervisor-id supervisor)
-                             " on port "
-                             port
-                             " with id "
-                             id
-                             )
-                (launch-worker supervisor
-                               (:storm-id assignment)
-                               port
-                               id
-                               mem-onheap)
-                (mark! supervisor:num-workers-launched)
-                (catch java.io.FileNotFoundException e
-                  (log-message "Unable to launch worker due to "
-                               (.getMessage e)))
-                (catch java.io.IOException e
-                  (log-message "Unable to launch worker due to "
-                               (.getMessage e))))
-         id)))
-    ))
+    (let [valid-new-worker-ids
+          (into {}
+            (remove nil?
+              (dofor [[port assignment] reassign-executors]
+                (let [id (new-worker-ids port)
+                      storm-id (:storm-id assignment)
+                      ^WorkerResources resources (:resources assignment)
+                      mem-onheap (.get_mem_on_heap resources)]
+                  ;; This condition checks for required files exist before launching the worker
+                  (if (required-topo-files-exist? conf storm-id)
+                    (do
+                      (log-message "Launching worker with assignment "
+                        (pr-str assignment)
+                        " for this supervisor "
+                        (:supervisor-id supervisor)
+                        " on port "
+                        port
+                        " with id "
+                        id)
+                      (local-mkdirs (worker-pids-root conf id))
+                      (local-mkdirs (worker-heartbeats-root conf id))
+                      (launch-worker supervisor
+                        (:storm-id assignment)
+                        port
+                        id
+                        mem-onheap)
+                      [port id])
```

Comment:
If we return `[id port]` here instead of `[port id]` here, we do not have to reverse the map later on when we use it.